Library/Homebrew/cmd/upgrade.rb

```diff
@@ -192,6 +192,9 @@ def upgrade_formula(f)
     end
     oh1 "Upgrading #{Formatter.identifier(f.full_specified_name)} #{upgrade_version} #{fi.options.to_a.join(" ")}"
 
+    # fetch and build dependencies, bottles, and non bottles
+    fi.fetch_and_build
```

Comment:
This will result in potentially building/upgrading dependencies so I'm not sure this quite gets us to where we want. Instead, check out the code for `cmd/fetch.rb`; anything that will be installed (e.g. `formulae_to_install` above) should be `fetch`ed first and then you can set in `formula_installer.rb` that it doesn't need refetched (it'll be a no-op i.e. not redownload but it will display the checksum etc. which we don't want).

Make sense?

Reply:
After this PR, do you want `build` to happen before or after unlinking?

Reply:
If it can be done before with no ill effects: that seems better. If not: no worries.

Reply:
@MikeMcQuaid 

I opted to focus only on fetching with this change. The `fetch` method inside `formula_installer` will now fetch only for both bottles and non-bottles. However, for non-bottles, there is an annoying display/output issue through the terminal.

See below: 

```
==> Downloading https://curl.haxx.se/download/curl-7.69.1.tar.bz2
Already downloaded: /Users/avocado/Library/Caches/Homebrew/downloads/be9bfe94edb7fc0d0c87d756311054a8c8dabc6d97cdb9a7c12c078a569d9b72--curl-7.69.1.tar.bz2
==> Downloading https://curl.haxx.se/download/curl-7.69.1.tar.bz2
Already downloaded: /Users/avocado/Library/Caches/Homebrew/downloads/be9bfe94edb7fc0d0c87d756311054a8c8dabc6d97cdb9a7c12c078a569d9b72--curl-7.69.1.tar.bz2
``` 
You can see that the downloaded is called twice resulting in the download output being duplicated.

Not the best user experience, but it isn't downloading twice. The issue is that fetch is being called twice, once inside the `fetch` method in `formula_installer` and again during `build`.

There isn't a simple way to track fetch status for non-bottles, as fetching happens within the downloaders and different instances of those can exist during `brew upgrade`.

I'd love some direction/suggestion on a way to fix this minor issue if you can provide any.

Reply:
Impressive work @MichelleNguy 👏. Have comment inline on what I think may help fix the problem (at least partially). If it doesn't completely I'll have another play around with it locally and a think. Thanks again!